workers/unity/Packages/com.improbable.gdk.deploymentlauncher/DeploymentLauncherWindow.cs

```diff
@@ -0,0 +1,598 @@
+            public readonly List<WrappedTask<Result<RedirectedProcessResult, Ipc.Error>, BaseDeploymentConfig>> CompletedLaunchTasks = new List<WrappedTask<Result<RedirectedProcessResult, Ipc.Error>, BaseDeploymentConfig>>();
+
+            private WrappedTask<RedirectedProcessResult, AssemblyConfig> uploadTask;
+            private WrappedTask<Result<RedirectedProcessResult, Ipc.Error>, BaseDeploymentConfig> launchTask;
+            private WrappedTask<Result<RedirectedProcessResult, Ipc.Error>, DeploymentInfo> stopTask;
+            private WrappedTask<Result<List<DeploymentInfo>, Ipc.Error>, string> listTask;
+
+            private Queue<(string, string, BaseDeploymentConfig)> queuedLaunches = new Queue<(string, string, BaseDeploymentConfig)>();
+
+            public void Upload(string projectName, AssemblyConfig config)
+            {
+                EditorApplication.LockReloadAssemblies();
+                IsUploading = true;
+                uploadTask = Assembly.UploadAsync(projectName, config);
+            }
+
+            public void Launch(string projectName, string assemblyName, BaseDeploymentConfig config)
+            {
+                if (IsLaunching)
+                {
+                    queuedLaunches.Enqueue((projectName, assemblyName, config));
+                    return;
+                }
+
+                EditorApplication.LockReloadAssemblies();
+                IsLaunching = true;
+                launchTask = Deployment.LaunchAsync(projectName, assemblyName, config);
+            }
+
+            public void Update()
+            {
+                if (uploadTask?.Task.IsCompleted == true)
+                {
+                    IsUploading = false;
+                    EditorApplication.UnlockReloadAssemblies();
+                    CompletedUploadTasks.Add(uploadTask);
+                    uploadTask = null;
+                }
+
+                if (launchTask?.Task.IsCompleted == true)
+                {
+                    CompletedLaunchTasks.Add(launchTask);
+
+                    if (queuedLaunches.Count > 0)
+                    {
+                        var (projectName, assemblyName, config) = queuedLaunches.Dequeue();
+                        launchTask = Deployment.LaunchAsync(projectName, assemblyName, config);
+                    }
+                    else
+                    {
+                        launchTask = null;
+                        IsLaunching = false;
+                        EditorApplication.UnlockReloadAssemblies();
+                    }
+                }
+            }
+
+            public string GetStatusMessage()
+            {
+                var sb = new StringBuilder();
+
+                if (IsUploading)
+                {
+                    sb.AppendLine($"Uploading assembly \"{uploadTask.Context.AssemblyName}\".");
```

Comment:
we should disable the launch deployment button as well. What happens if I start a deployment just after starting the upload for t he assembly. It seems to start fine, but I have no idea what the behaviour will be.

Reply:
It would be a race most likely, but yes we should probably do that as well.